[css-values] Implement rcap / rex / ric / rch units

nt1m opened the pull request:
#### 424dee4815091e03fcc4c9035af6e7a9a8128184
<pre>
[css-values] Implement rcap / rex / ric / rch units
<a href="https://bugs.webkit.org/show_bug.cgi?id=260648">https://bugs.webkit.org/show_bug.cgi?id=260648</a>
rdar://114367871

Reviewed by Cameron McCormack.

Implement new units as per <a href="https://drafts.csswg.org/css-values/#font-relative-lengths">https://drafts.csswg.org/css-values/#font-relative-lengths</a>

Invalidation support is also added by invalidating when the root font cascade / line height changes.

Also add a bunch of tests for root font units in monospace.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rcap-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rch-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rex-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/ric-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rlh-invalidation-expected.txt:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::computeUnzoomedNonCalcLengthDouble):
(WebCore::CSSPrimitiveValue::computeNonCalcLengthDouble):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::CSSPrimitiveValue::isRootFontRelativeLength):
(WebCore::CSSPrimitiveValue::isFontRelativeLength):
(WebCore::CSSPrimitiveValue::isLength):
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator&lt;&lt;):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::calcUnitCategory):
(WebCore::calculationCategoryForCombination):
(WebCore::hasDoubleValue):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::cssPrimitiveValueUnitFromTrie):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::LengthRawKnownTokenTypeDimensionConsumer::consume):
* Source/WebCore/css/typedom/CSSNumericFactory.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::styleChangeAffectsRelativeUnits):

Canonical link: <a href="https://commits.webkit.org/267321@main">https://commits.webkit.org/267321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fd5fb2050a30949b7b37bebc11c7b8ec3f5139f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16293 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17031 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18061 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15280 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19693 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16725 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17661 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16487 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13952 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18828 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/14749 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21564 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15159 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/14914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18241 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15507 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14590 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/3893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19091 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1999 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15334 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace.html:
```diff
@@ -0,0 +1,16 @@
+
+<div style="font-family: sans-serif; font-size: 1ric">
+  Text.
+</div>
```
Can we make this use a font that has a 水 glyph, like the ic-unit-* tests, and have it on the root element? I feel like we're missing out on some test coverage by not having ric reference such a font on the root element.

And similarly for the rex test?

These tests feel like they're just checking that ric / rex compute against something other than the parent, but it would be good to check that they're computing against exactly what we want.

nt1m replied:
Done!

heycam reviewed (approved):
r=me with some test changes.

nt1m committed:
[css-values] Implement rcap / rex / ric / rch units
https://bugs.webkit.org/show_bug.cgi?id=260648
rdar://114367871

Reviewed by Cameron McCormack.

Implement new units as per https://drafts.csswg.org/css-values/#font-relative-lengths

Invalidation support is also added by invalidating when the root font cascade / line height changes.

Also add a bunch of tests for root font units in monospace.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rcap-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rch-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/rex-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/ric-in-monospace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rcap-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rch-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rex-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/ric-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/rlh-invalidation-expected.txt:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::isValidCSSUnitTypeForDoubleConversion):
(WebCore::isStringType):
(WebCore::CSSPrimitiveValue::~CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::computeUnzoomedNonCalcLengthDouble):
(WebCore::CSSPrimitiveValue::computeNonCalcLengthDouble):
(WebCore::CSSPrimitiveValue::unitTypeString):
(WebCore::CSSPrimitiveValue::serializeInternal const):
(WebCore::CSSPrimitiveValue::equals const):
(WebCore::CSSPrimitiveValue::addDerivedHash const):
(WebCore::CSSPrimitiveValue::collectComputedStyleDependencies const):
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::CSSPrimitiveValue::isRootFontRelativeLength):
(WebCore::CSSPrimitiveValue::isFontRelativeLength):
(WebCore::CSSPrimitiveValue::isLength):
* Source/WebCore/css/CSSUnits.cpp:
(WebCore::unitCategory):
(WebCore::operator<<):
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp:
(WebCore::calcUnitCategory):
(WebCore::calculationCategoryForCombination):
(WebCore::hasDoubleValue):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::cssPrimitiveValueUnitFromTrie):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::LengthRawKnownTokenTypeDimensionConsumer::consume):
* Source/WebCore/css/typedom/CSSNumericFactory.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
(WebCore::CSSStyleValueFactory::reifyValue):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::styleChangeAffectsRelativeUnits):

Canonical link: https://commits.webkit.org/267321@main
